File: apps/macos-native/G9Claw/Services/CodeSyntaxHighlightingService.swift
import Foundation
import HighlightSwift
import SwiftUI

enum CodeSyntaxHighlightingService {
    static let maxHighlightedCharacters = 180_000
    static let highlightDebounceInterval: TimeInterval = 0.18

    static func languageAlias(forFileName fileName: String) -> String? {
        let lowercased = fileName.lowercased()
        let ext = (lowercased as NSString).pathExtension
        let baseName = (lowercased as NSString).lastPathComponent

        switch baseName {
        case "dockerfile":
            return "dockerfile"
        case "makefile":
            return "makefile"
        default:
            break
        }

        switch ext {
        case "html", "htm":
            return "html"
        case "css":
            return "css"
        case "scss":
            return "scss"
        case "js", "mjs", "cjs":
            return "javascript"
        case "ts", "tsx":
            return "typescript"
        case "jsx":
            return "javascript"
        case "py", "pyw":
            return "python"
        case "swift":
            return "swift"
        case "json", "jsonc":
            return "json"
        case "yml", "yaml":
            return "yaml"
        case "md", "markdown":
            return "markdown"
        case "sh", "bash", "zsh":
            return "bash"
        case "toml":
            return "toml"
        case "xml":
            return "xml"
        case "sql":
            return "sql"
        case "rs":
            return "rust"
        case "go":
            return "go"
        case "java":
            return "java"
        case "c":
            return "c"
        case "cc", "cpp", "cxx", "hpp", "hh":
            return "cpp"
        default:
            return nil
        }
    }

    static func shouldHighlight(_ text: String, languageAlias: String?) -> Bool {
        languageAlias != nil && !text.isEmpty && text.count <= maxHighlightedCharacters
    }

    static func themeName(isDarkMode: Bool) -> String {
        isDarkMode ? "tokyoNight" : "xcode"
    }

    static func colors(isDarkMode: Bool) -> HighlightColors {
        isDarkMode ? .dark(.tokyoNight) : .light(.xcode)
    }

    static func highlightedText(
        _ text: String,
        languageAlias: String,
        isDarkMode: Bool
    ) async throws -> AttributedString {
        try await Highlight().attributedText(
            text,
            language: languageAlias,
            colors: colors(isDarkMode: isDarkMode)
        )
    }
}

File: apps/windows/src/G9Claw.Windows.Core/Services/CodeSyntaxHighlightingService.cs
namespace G9Claw.Windows.Core;

public static class CodeSyntaxHighlightingService
{
    public const int MaxHighlightedCharacters = 180_000;
    public static readonly TimeSpan HighlightDebounceInterval = TimeSpan.FromMilliseconds(180);

    public static string? LanguageAliasForFileName(string fileName)
    {
        var lower = fileName.Trim().ToLowerInvariant();
        var baseName = Path.GetFileName(lower);
        var extension = Path.GetExtension(lower).TrimStart('.');

        switch (baseName)
        {
            case "dockerfile":
                return "dockerfile";
            case "makefile":
                return "makefile";
        }

        return extension switch
        {
            "html" or "htm" => "html",
            "css" => "css",
            "scss" => "scss",
            "js" or "mjs" or "cjs" or "jsx" => "javascript",
            "ts" or "tsx" => "typescript",
            "py" or "pyw" => "python",
            "swift" => "swift",
            "json" or "jsonc" => "json",
            "yml" or "yaml" => "yaml",
            "md" or "markdown" => "markdown",
            "sh" or "bash" or "zsh" => "bash",
            "toml" => "toml",
            "xml" => "xml",
            "sql" => "sql",
            "rs" => "rust",
            "go" => "go",
            "java" => "java",
            "c" => "c",
            "cc" or "cpp" or "cxx" or "hpp" or "hh" => "cpp",
            _ => null,
        };
    }

    public static bool ShouldHighlight(string text, string? languageAlias) =>
        !string.IsNullOrWhiteSpace(languageAlias) &&
        text.Length > 0 &&
        text.Length <= MaxHighlightedCharacters;

    public static string ThemeName(bool isDarkMode) => isDarkMode ? "tokyoNight" : "xcode";
}
